lib/min_heap.rb

```diff
@@ -55,17 +61,46 @@ def empty?
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: o(1)
+  # Space complexity: o(1)
   def heap_up(index)
-    
+    return if index == 0
+    parent_index = (index % 2 != 0) ? (index / 2) : (index / 2 - 1)
+    if @store[index].key < @store[parent_index].key
+      swap(index,parent_index)
+      heap_up(parent_index)
+    else
+      return
+    end
   end
 
   # This helper method takes an index and 
-  #  moves it up the heap if it's smaller
-  #  than it's parent node.
+  #  moves it down the heap if it's bigger
+  #  than its child node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    return if index == nil
+    left_child_index = (((index * 2) + 1) < (@store.length - 1)) ? ((index * 2) + 1) : nil
+    right_child_index = (((index * 2) + 2) < (@store.length - 1)) ? ((index * 2) + 2) : nil
+
+    if !left_child_index && !right_child_index
+      return
+    elsif left_child_index && !right_child_index
+      if @store[index].key > @store[left_child_index].key
+        swap(index,left_child_index)
+      end
+    elsif !left_child_index && right_child_index
+      if @store[index].key > @store[right_child_index].key
+        swap(index,right_child_index)
+      end
+    else
+      if @store[left_child_index].key < @store[right_child_index].key
+        index_with_smaller_key = left_child_index
+      else
+        index_with_smaller_key = right_child_index
+      end
+      swap(index, index_with_smaller_key)
+    end
+    heap_down(index_with_smaller_key)  
```

Comment:
This might lead you to heaping down, even when you didn't do a swap... It won't cause bugs, but might lead to extra recursive calls.